server/src/main/java/org/opensearch/gateway/PrimaryShardBatchAllocator.java

```diff
@@ -0,0 +1,152 @@
+            if (decision != null) {
+                inEligibleShards.add(shard);
+                shardAllocationDecisions.put(shard, decision);
+            } else {
+                eligibleShards.add(shard);
+            }
+        }
+        // Do not call fetchData if there are no eligible shards
+        if (eligibleShards.isEmpty()) {
+            return shardAllocationDecisions;
+        }
+        // only fetch data for eligible shards
+        final FetchResult<NodeGatewayStartedShardsBatch> shardsState = fetchData(eligibleShards, inEligibleShards, allocation);
+
+        // process the received data
+        for (ShardRouting unassignedShard : eligibleShards) {
+            List<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> nodeShardStates = adaptToNodeShardStates(
+                unassignedShard,
+                shardsState
+            );
+            // get allocation decision for this shard
+            shardAllocationDecisions.put(unassignedShard, getAllocationDecision(unassignedShard, allocation, nodeShardStates, logger));
+        }
+        return shardAllocationDecisions;
+    }
+
+    /**
+     * Transforms {@link FetchResult} of {@link NodeGatewayStartedShardsBatch} to {@link List} of {@link TransportNodesListGatewayStartedShards.NodeGatewayStartedShards}.
+     * <p>
+     * Returns null if {@link FetchResult} does not have any data.
+     * <p>
+     * shardsState contain the Data, there key is DiscoveryNode but value is Map of ShardId
+     * and NodeGatewayStartedShardsBatch so to get one shard level data (from all the nodes), we'll traverse the map
+     * and construct the nodeShardState along the way before making any allocation decision. As metadata for a
+     * particular shard is needed from all the discovery nodes.
+     *
+     * @param unassignedShard unassigned shard
+     * @param shardsState fetch data result for the whole batch
+     * @return shard state returned from each node
+     */
+    private static List<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> adaptToNodeShardStates(
+        ShardRouting unassignedShard,
+        FetchResult<NodeGatewayStartedShardsBatch> shardsState
+    ) {
+        if (!shardsState.hasData()) {
+            return null;
+        }
+        List<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> nodeShardStates = new ArrayList<>();
+        Map<DiscoveryNode, NodeGatewayStartedShardsBatch> nodeResponses = shardsState.getData();
+
+        // build data for a shard from all the nodes
+        nodeResponses.forEach((node, nodeGatewayStartedShardsBatch) -> {
+            NodeGatewayStartedShard shardData = nodeGatewayStartedShardsBatch.getNodeGatewayStartedShardsBatch()
+                .get(unassignedShard.shardId());
+            nodeShardStates.add(
+                new TransportNodesListGatewayStartedShards.NodeGatewayStartedShards(
```

Comment:
This was done already as part of refactoring https://github.com/opensearch-project/OpenSearch/pull/9760/commits/4bf3949dd004dc79842b47d21d3fcf5f7323401f#diff-b6c6ef5bc5371b50e65de279de18409a16efd32d5e9d5ed1f6989406bb4a920eL548-L558 

but removed later after this discussion : https://github.com/opensearch-project/OpenSearch/pull/9760#discussion_r1418504715

Reason was to avoid touching existing `getAllocationDecision`.

Reply:
I've created a new `GatewayShardStarted` in helper class. That contains only 4 fields without DiscoveryNode. 
And another class `NodeGatewayShardStarted` which has extra DiscoveryNode object also, -> this'll be used by **PrimaryShardAllocator** to get the allocation decision. And `PrimaryShardBatchAllocator` will convert its FetchResult into `NodeGatewayShardStarted` (of helper class) to make use of `PrimaryShardAllocator` **getAllocationDecision** method.